web/packages/teleport/src/services/api/api.js

```diff
@@ -78,16 +90,27 @@ const api = {
           } else {
             return response
               .json()
-              .then(json => reject(new ApiError(parseError(json), response)))
-              .catch(err => {
+              .then(json => {
+                if (
+                  !withMFA &&
+                  isAdminActionRequiresMFAError(parseError(json))
+                ) {
+                  // Retry with MFA.
+                  return this.fetchJson(url, params, true)
+                    .then(resp => resolve(resp))
+                    .catch(err => reject(err));
```

Comment:
Perhaps we could have only have `return this.fetchJson(url, params, true);`? Since it returns a promise and other callbacks what handle it (but I didn't test it).

Overall, this code would probably benefit a lot from rewriting to async/await, it's so hard to read :/ But I don't require doing it :)

Reply:
Agreed, I don't mind the rewriting exercise as I'm still trying to get a grasp on it. Will update.